### ISSUE TITLE
Segmentation fault in Transform process

### ISSUE THREAD

**Issue by pratikshetty1994** [dimitri/pgcopydb#679]
Context:
Moving from aws rds postgres 14.3 to aws rds postygres 14.5.
Command used: pgcopy clone --follow  --split-tables-larger-than 1GB --table-jobs 1 --index-jobs 1

We are getting two issues not sure if they are co-related:
1) database lock in the initial state of the process. It occurs only once.
05:29:44.038 10 ERROR  Failed to execute statement: update sentinel set startpos = $1, write_lsn = $2, flush_lsn = $3 where id = 1
05:29:44.038 10 ERROR  [SQLite 5] database is locked

2) Segementation fault in transform process and follow process exiting.
06:33:14.926 11 NOTICE Transforming JSON file "/dbhome/tmp//cdc/00000001000072B60000002F.json" into SQL file "/dbhome/tmp//cdc/00000001000072B60000002F.sql"
06:33:15.492 9 ERROR  Subprocess transform with pid 11 has exited with return code 0 and signal Segmentation fault
06:33:15.492 9 NOTICE Process transform has exited unexpectedly, and endpos is unset: terminating other processes
06:33:15.492 9 NOTICE kill -TERM 10 (prefetch)
06:33:15.492 9 NOTICE kill -TERM 12 (catchup)
06:33:15.492 12 INFO   Apply process received a shutdown signal while waiting for apply mode, quitting now
06:33:15.492 12 NOTICE Apply mode is still disabled, quitting now
06:33:15.492 12 INFO   Apply process has terminated
06:33:15.495 10 INFO   Reported write_lsn 72B6/C00DB978, flush_lsn 72B6/C00DB978, replay_lsn 0/0
06:33:15.496 10 NOTICE Closed file "/dbhome/tmp//cdc/00000001000072B600000030.json"
06:33:15.496 10 INFO   Streamed up to write_lsn 72B6/C00DB978, flush_lsn 72B6/C00DB978, stopping: endpos is 0/0
06:33:15.496 10 INFO   Prefetch process has terminated
06:33:15.644 9 NOTICE Subprocess prefetch with pid 10 has exited successfully
06:33:15.644 9 NOTICE Process prefetch has exited successfully, and endpos is unset: terminating other processes
06:33:15.644 9 NOTICE kill -TERM 12 (catchup)
06:33:15.644 9 NOTICE Subprocess catchup with pid 12 has exited successfully
06:33:15.644 9 NOTICE Process catchup has exited successfully, and endpos is unset: terminating other processes
06:33:15.795 9 ERROR  Some sub-process exited with errors, see above for details
06:33:15.795 9 ERROR  Failed to follow changes from source, see above for details

Can you please help us identify what the issue could be?

**Comment by pratikshetty1994** [comment]
```(gdb) bt full
#0  catalog_s_table_attrs_fetch (query=0x7fa0e2daf920) at catalog.c:3351
        table = 0x7fa0e2db7800
        attnum = 2
        count = <optimized out>
        attr = 0x8c
#1  0x000055f2f2af0d08 in catalog_s_table_fetch_attrs (catalog=catalog@entry=0x7ffd674700f0, table=table@entry=0x7fa0e2db7800) at catalog.c:3202
        iter = 0x7fa0e2daf910
#2  0x000055f2f2b1600d in prepareUpdateTuppleArrays (header=0x7ffd672518c0, privateContext=0x7ffd674700f0) at ld_test_decoding.c:983
        cols = 0x7fa0e2db6c30
        table = 0x7fa0e2db7800
        pkeyArray = <optimized out>
        newCount = <optimized out>
        newPos = <optimized out>
        stmt = 0x7fa0e2db33c0
        columnCount = <optimized out>
        new = <optimized out>
        oldPos = <optimized out>
        sourceDB = 0x7ffd674700f0
        oldCount = <optimized out>
        old = <optimized out>
        stmt = <optimized out>
        cols = <optimized out>
        sourceDB = <optimized out>
        table = <optimized out>
        columnCount = <optimized out>
        pkeyArray = <optimized out>
        oldCount = <optimized out>
        newCount = <optimized out>
        old = <optimized out>
        new = <optimized out>
        oldPos = <optimized out>
        newPos = <optimized out>
        c = <optimized out>
        colname = <optimized out>
        attribute = <optimized out>
        c = <optimized out>
        colname = <optimized out>
#3  parseTestDecodingUpdateMessage (header=0x7ffd672518c0, privateContext=0x7ffd674700f0) at ld_test_decoding.c:505
        stmt = <optimized out>
        stmt = <optimized out>
        old = <optimized out>
        new = <optimized out>
#4  parseTestDecodingMessage (privateContext=privateContext@entry=0x7ffd6735c4e0,
    message=message@entry=0x7fa0e3106cde "{\"action\":\"U\",\"xid\":\"0\",\"lsn\":\"7344/E62C9148\",\"timestamp\":\"2024-02-19 23:04:40.888832+0000\",\"message\":\"table public.idemia_request: UPDATE: card[bigint]:594191474 cardholder[bigint]:225540831 serial_n"..., json=json@entry=0x7fa0e2db6e10) at ld_test_decoding.c:270
        stmt = <optimized out>
        metadata = 0x7ffd6735e180
        jsobj = <optimized out>
--Type <RET> for more, q to quit, c to continue without paging--RET
        header = {
          message = 0x7fa0e2db57e0 "table public.idemia_request: UPDATE: card[bigint]:594191474 cardholder[bigint]:225540831 serial_no[character varying]:'000009' file_name[character varying]:'fpl_embosser_SBM_P0101_BLUEDART_IDEMIA_1707"..., qname = "public.idemia_request", '\000' <repeats 239 times>, table = {nspname = 0x55f404e155c0 "public",
            relname = 0x55f404e155e0 "idemia_request", pqMemory = false}, action = STREAM_ACTION_UPDATE, offset = 37, pos = 526, eom = true}
        td_message = <optimized out>
#5  0x000055f2f2b17542 in parseMessage (privateContext=privateContext@entry=0x7ffd6735c4e0,
    message=message@entry=0x7fa0e3106cde "{\"action\":\"U\",\"xid\":\"0\",\"lsn\":\"7344/E62C9148\",\"timestamp\":\"2024-02-19 23:04:40.888832+0000\",\"message\":\"table public.idemia_request: UPDATE: card[bigint]:594191474 cardholder[bigint]:225540831 serial_n"..., json=0x7fa0e2db6e10) at ld_transform.c:1305
        jsmesgtype = <optimized out>
        mesg = 0x7ffd6735e118
        metadata = 0x7ffd6735e180
        txn = <optimized out>
        stmt = 0x7fa0e2db33c0
#6  0x000055f2f2b1986b in stream_transform_file (specs=specs@entry=0x7ffd67253210, jsonfilename=jsonfilename@entry=0x7ffd672523b0 "/dbhome/tmp//cdc/000000010000734400000039.json",
    sqlfilename=sqlfilename@entry=0x7ffd672527b0 "/dbhome/tmp//cdc/000000010000734400000039.sql") at ld_transform.c:995
        message = 0x7fa0e3106cde "{\"action\":\"U\",\"xid\":\"0\",\"lsn\":\"7344/E62C9148\",\"timestamp\":\"2024-02-19 23:04:40.888832+0000\",\"message\":\"table public.idemia_request: UPDATE: card[bigint]:594191474 cardholder[bigint]:225540831 serial_n"...
        empty = <optimized out>
        json = 0x7fa0e2db6e10
        i = 127
        privateContext = 0x7ffd6735c4e0
        content = {filename = "/dbhome/tmp//cdc/000000010000734400000039.json", '\000' <repeats 977 times>, lbuf = {
            buffer = 0x7fa0e30fa000 "{\"action\":\"B\",\"xid\":\"2538869762\",\"lsn\":\"7344/E40DF1A0\",\"timestamp\":\"2024-02-19 23:04:15.62934+0000\",\"message\":\"BEGIN 2538869762\"}", count = 220, lines = 0x7fa0e3110800}, messages = 0x0}
        contents = 0x7fa0e30fa000 "{\"action\":\"B\",\"xid\":\"2538869762\",\"lsn\":\"7344/E40DF1A0\",\"timestamp\":\"2024-02-19 23:04:15.62934+0000\",\"message\":\"BEGIN 2538869762\"}"
        size = 89442
        tempfilename = "/dbhome/tmp//cdc/000000010000734400000039.sql.partial", '\000' <repeats 970 times>
        currentMsgIndex = 32
        currentMsg = 0x7ffd6735e118
        metadata = 0x7ffd6735e180
        firstMessage = false
#7  0x000055f2f2b1a094 in stream_transform_file_at_lsn (specs=specs@entry=0x7ffd67253210, lsn=<optimized out>) at ld_transform.c:789
        walFileName = "/dbhome/tmp//cdc/000000010000734400000039.json", '\000' <repeats 977 times>
        sqlFileName = "/dbhome/tmp//cdc/000000010000734400000039.sql", '\000' <repeats 978 times>
#8  0x000055f2f2b1a24e in stream_transform_from_queue_internal (specs=0x7ffd67253210) at ld_transform.c:733
        mesg = {type = 4, data = {oid = 3826119072, lsn = 126739721089440, tp = {oid = 3826119072, part = 29508}}}
        recv_ok = <optimized out>
        transformQueue = 0x7ffd6735bf90
        errors = <optimized out>
        stop = false
        success = <optimized out>
        transformQueue = <optimized out>
        errors = <optimized out>
        stop = <optimized out>
        success = <optimized out>
        mesg = <optimized out>
        recv_ok = <optimized out>
--Type <RET> for more, q to quit, c to continue without paging--
#9  stream_transform_from_queue (specs=specs@entry=0x7ffd67253210) at ld_transform.c:671
        sourceDB = 0x7ffd674700f0
        success = <optimized out>
#10 0x000055f2f2b1a385 in stream_transform_worker (specs=specs@entry=0x7ffd67253210) at ld_transform.c:645
No locals.
#11 0x000055f2f2b0a45e in follow_start_transform (specs=0x7ffd67253210) at follow.c:768
        success = <optimized out>
#12 0x000055f2f2b0a767 in follow_start_subprocess (specs=specs@entry=0x7ffd67253210, subprocess=subprocess@entry=0x7ffd6735bf48) at follow.c:896
        pid = <optimized out>
        psTitle = "pgcopydb: follow transform", '\000' <repeats 997 times>
        ps = {pid = 12, psType = "transform", '\000' <repeats 120 times>, psTitle = 0x7ffd67475daa "pgcopydb: follow transform", tableOid = 0, partNumber = 0, indexOid = 0}
        sourceDB = 0x7ffd674700f0
        fpid = <optimized out>
#13 0x000055f2f2b0b43b in followDB (copySpecs=<optimized out>, streamSpecs=0x7ffd67253210) at follow.c:586
        errors = <optimized out>
        prefetch = 0x7ffd6735bf28
        transform = 0x7ffd6735bf48
        catchup = 0x7ffd6735bf68
        errors = <optimized out>
        prefetch = <optimized out>
        transform = <optimized out>
        catchup = <optimized out>
#14 followDB (copySpecs=<optimized out>, streamSpecs=0x7ffd67253210) at follow.c:509
        errors = 0
        prefetch = <optimized out>
        transform = <optimized out>
        catchup = <optimized out>
#15 0x000055f2f2b0b76b in follow_main_loop (copySpecs=copySpecs@entry=0x7ffd6735ef00, streamSpecs=streamSpecs@entry=0x7ffd67253210) at follow.c:290
        done = false
        sourceDB = 0x7ffd674700f0
        modeArray = {STREAM_MODE_CATCHUP, STREAM_MODE_REPLAY}
        count = 2
        loop = 0
        currentMode = STREAM_MODE_CATCHUP
        previousMode = <optimized out>
#16 0x000055f2f2af5176 in start_follow_process (pid=<synthetic pointer>, streamSpecs=0x7ffd67253210, copySpecs=0x7ffd6735ef00) at cli_clone_follow.c:683
        sentinel = 0x7ffd6735bef0
        fpid = 0
        sentinel = <optimized out>
        fpid = <optimized out>
#17 clone_and_follow (copySpecs=0x7ffd6735ef00) at cli_clone_follow.c:264
        logSQL = <optimized out>
        clonePID = 9
        success = <optimized out>
        streamSpecs = <error reading variable streamSpecs (value of type `StreamSpecs' requires 1096936 bytes, which is more than max-value-size)>
        followPID = -1
        streamSpecs = <optimized out>
        logSQL = <optimized out>
--Type <RET> for more, q to quit, c to continue without paging--c
        clonePID = <optimized out>
        followPID = <optimized out>
        success = <optimized out>
#18 cli_clone (argc=<optimized out>, argv=<optimized out>) at cli_clone_follow.c:137
        copySpecs = <error reading variable copySpecs (value of type `CopyDataSpec' requires 1137736 bytes, which is more than max-value-size)>
        exportSnapshot = <optimized out>
        clonePID = 0
        success = <optimized out>
        allExitsAreZero = <optimized out>
#19 0x000055f2f2b3ba3a in commandline_run (command=command@entry=0x7ffd67474bd0, argc=0, argc@entry=19, argv=0x7ffd67474de0, argv@entry=0x7ffd67474d48)
    at /usr/src/pgcopydb/src/bin/pgcopydb/../lib/subcommands.c/commandline.c:71
        argv0 = <optimized out>
        breadcrumb = 0x55f2f4b91a20 "pgcopydb clone"
#20 0x000055f2f2ae8ec1 in main (argc=19, argv=0x7ffd67474d48) at main.c:140
        command = {name = 0x55f2f2c28b78 "pgcopydb", shortDescription = 0x55f2f2c25ddb "pgcopydb tool", usageSuffix = 0x55f2f2c25de9 "[ --verbose --quiet ]", help = 0x0,
          getopt = 0x55f2f2afcfe0 <root_options>, run = 0x0, subcommands = 0x55f2f2c7e260 <root_subcommands>, breadcrumb = 0x0}```

**Comment by pratikshetty1994** [comment]
coredump for context

**Comment by dimitri** [comment]
Which version of pgcopydb are you using? Could you try the most recent version, locally built from current sources?

**Comment by pratikshetty1994** [comment]
@ it was locally built from current sources. I think we figured out the issue.. We have a table whose attnum starts from 2 instead of 1. We fixed it with a makeshift solution by adding the below code in catalog_s_table_attrs_fetch function in catalog.c
```
if (oid == 4118319 && attnum == 2)
	{
		log_debug("Inside Custom If Condition");
		log_debug("Table oid: %d", oid);
		table->attributes.count = count;
		table->attributes.array =
			(SourceTableAttribute *) calloc(count,
											sizeof(SourceTableAttribute));

		if (table->attributes.array == NULL)
		{
			log_error(ALLOCATION_FAILED_ERROR);
			return false;
		}
	}
```

**Comment by pratikshetty1994** [comment]
@dimitri  but now we are getting the following error. it does not seem related. The debug log does print it as a BUG. Do you know if the error below is due to an issue with the test_decoding.
```
13:14:04.874 11 ERROR  ld_apply.c:1766           BUG: computeTxnMetadataFilename is called with transaction xid: 0
13:14:04.874 11 ERROR  ld_apply.c:744            Failed to write transaction metadata file, see above for details
13:14:04.874 11 ERROR  ld_apply.c:523            Failed to apply SQL from file "/dbhome/tmp//cdc/000000010000747B00000006.sql", see above for details
```

**Comment by dimitri** [comment]
> @dimitri it was locally built from current sources. I think we figured out the issue.. We have a table whose attnum starts from 2 instead of 1. We fixed it with a makeshift solution by adding the below code in catalog_s_table_attrs_fetch function in catalog.c

Good catch thanks! I have now merged a fix to attnum not being a gapless series that starts at 1 by using our own computed numbering, which is easy to obtain when we have window-functions in SQLite. Marked as fixed.

About the test_decoding problem, if that problem persists, please consider opening a separate issue for it.
